src/main/scala/firrtl/transforms/FixAddingNegativeLiteralsTransform.scala

```diff
@@ -0,0 +1,106 @@
+// See LICENSE for license details.
+
+package firrtl.transforms
+
+import firrtl.{CircuitState, LowForm, Namespace, PrimOps, Transform, Utils, WRef}
+import firrtl.ir._
+import firrtl.Mappers._
+import firrtl.PrimOps.{Add, AsSInt, Sub, Tail}
+
+import scala.collection.mutable
+
+object FixAddingNegativeLiterals {
+
+  /** Returns the maximum negative number represented by given width
+    * @param width width of the negative number
+    * @return maximum negative number
+    */
+  def maxNegNumber(width: BigInt): BigInt = -(1 << (width.toInt - 1))
+
+  /** Updates the type of the DoPrim from its arguments (e.g. if is UnknownType)
+    * @param d input DoPrim
+    * @return updated DoPrim with calculated type
+    */
+  def setType(d: DoPrim): DoPrim = {
+    PrimOps.set_primop_type(d)
+  }
+
+  /** Returns a module with fixed additions of negative literals
+    * @param m input module
+    * @return updated module
+    */
+  def fixupModule(m: DefModule): DefModule = {
+    val namespace = Namespace(m)
+    m map fixupStatement(namespace)
+  }
+
+  /** Returns a statement with fixed additions of negative literals
+    * @param namespace object to enabling creating unique names
+    * @param s input statement
+    * @return updated statement
+    */
+  def fixupStatement(namespace: Namespace)(s: Statement): Statement = {
+    val stmtBuffer = mutable.ArrayBuffer[Statement]()
+    val ret = s map fixupStatement(namespace) map fixupExpression(Utils.get_info(s), namespace, stmtBuffer)
+    if(stmtBuffer.isEmpty) {
+      ret
+    } else {
+      stmtBuffer += ret
+      Block(stmtBuffer.toList)
```

Comment:
If you're going to cast to a List anyway, you should use `ListBuffer` instead of `ArrayBuffer` because casting to a `List` **is free** for ListBuffers (it copies the `ListBuffer` on write if you continue mutating).